services/snapshotter/service.go

```diff
@@ -274,12 +359,46 @@ func (s *Service) writeRetentionPolicyInfo(conn net.Conn, database, retentionPol
 }
 
 // readRequest unmarshals a request object from the conn.
-func (s *Service) readRequest(conn net.Conn) (Request, error) {
+func (s *Service) readRequest(conn net.Conn) (Request, []byte, error) {
 	var r Request
-	if err := json.NewDecoder(conn).Decode(&r); err != nil {
-		return r, err
+	d := json.NewDecoder(conn)
+
+	if err := d.Decode(&r); err != nil {
+		return r, []byte{}, err
 	}
-	return r, nil
+
+	bits := make([]byte, r.UploadSize+1)
+
+	if r.UploadSize > 0 {
+
+		remainder := d.Buffered()
+
+		//// no time to investigate.  It seems the JSON encoder puts one extra byte on the stream,
+		//// but the decoder does not consume it.  so we eat a byte here and then everything works.
+		//onebyte := bufio.NewReader(remainder)
+		//_, err := onebyte.ReadByte()
+		//if err != nil {
+		//	return r, bits, err
+		//}
+
+		n, err := remainder.Read(bits)
+		if err != nil && err != io.EOF {
+			return r, bits, err
+		}
+
+		// it is a bit random but sometimes the Json decoder will consume all the bytes and sometimes
+		// it will leave a few behind.
+		if err != io.EOF && n < int(r.UploadSize+1) {
+			n, err = conn.Read(bits[n:])
```

Comment:
In general, prefer `ioutil.ReadAll` or `io.ReadFull` to ensure draining of readers or reading sufficient data.